Adds API to GEP-91: Client Certificate Validation (#2273)
kubernetes-sigs/gateway-api: geps/gep-91/index.md

    // the Certificate Authorities that can be used
    // as a trust anchor to validate the certificates presented by the client.
    //
    // A single CA certificate reference to a Kubernetes ConfigMap kind has "Core" support.
    // Implementations MAY choose to support attaching multiple CA certificates to
    // a Listener, but this behavior is implementation-specific.
    //
    // Support: Core - An optional single reference to a Kubernetes Secret and ConfigMap,
    // with the CA certificate in a key named `ca.crt`.
    //
    // Support: Implementation-specific (More than one reference, or other kinds
    // of resources).
    //
    // References to a resource in a different namespace are invalid UNLESS there
    // is a ReferenceGrant in the target namespace that allows the certificate
    // to be attached. If a ReferenceGrant does not allow this reference, the
    // "ResolvedRefs" condition MUST be set to False for this listener with the
    // "RefNotPermitted" reason.
    //
    // +kubebuilder:validation:MaxItems=8
    // +optional
    CACertificateRefs []SecretObjectReference `json:”caCertificateRefs,omitempty”`

> }

Reviewer: Can we add SAN validation here? TLS without SAN verification isn't so good.

```suggestion

    // SubjectAltNames contains one or more alternate names to verify
    // the subject identity in the certificate presented by the client.
    //
    // Support: Core
    //
    // +optional
    SubjectAltNames []string `json:”subjectAltNames,omitempty”`
}
```

Author: ive added SAN as a non goal to this GEP, only because it may delay the graduation of this GEP. Happy to add this diff if it doesnt

Reviewer: I see the reverse problem wrt delay. If we don't add it here, it will delay the graduation of full/proper support for mTLS, because of all the extra overhead of starting up a second followup GEP to add SAN.

Author: cool, I'll add it, lets see how it goes ....

Author: added

to the Gateway.
* Introduce a `caCertificateRefs` field within `ClientValidationContext` that can be used to specify a list of CA Certificates that
can be used as a trust anchor to validate the certificates presented by the client.
* This new field is mutually exclusive with the [BackendTLSPolicy][] configuation which is used to validate the TLS certificate presented by the peer on the connection between the Gateway and the backend, and this GEP is adding support for validating the TLS certificate presented by the peer on the connection between the Gateway and the downstream client.

#### GO

```go
// ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection
// to the Gateway.
// By default, no client specific configuration is validated.
type ClientValidationContext struct {
    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of
    // the Certificate Authorities that can be used
    // as a trust anchor to validate the certificates presented by the client.
    //

>     // A single CA certificate reference to a Kubernetes ConfigMap kind has "Core" support.

Reviewer: ```suggestion
    // A single CA certificate reference to a Kubernetes ConfigMap or Secret kind
    // has "Core" support.
```

    // Implementations MAY choose to support attaching multiple CA certificates to
    // a Listener, but this behavior is implementation-specific.
    //

>     // Support: Core - An optional single reference to a Kubernetes Secret and ConfigMap,
>     // with the CA certificate in a key named `ca.crt`.

Reviewer: ```suggestion
    // Support: Core - An optional references to a single Kubernetes Secret and to a
    // single Kuberentes ConfigMap, with the CA certificate in a key named `ca.crt`.
```

    //
    // Support: Implementation-specific (More than one reference, or other kinds
    // of resources).
    //
    // References to a resource in a different namespace are invalid UNLESS there
    // is a ReferenceGrant in the target namespace that allows the certificate
    // to be attached. If a ReferenceGrant does not allow this reference, the
    // "ResolvedRefs" condition MUST be set to False for this listener with the
    // "RefNotPermitted" reason.
    //
    // +kubebuilder:validation:MaxItems=8
    // +kubebuilder:validation:MinItems=1
    CACertificateRefs []SecretObjectReference `json:”caCertificateRefs,omitempty”`
    // SubjectAltNames contains one or more alternate names to verify
    // the subject identity in the certificate presented by the client.
    //
    // Support: Core
    //
    // +optional

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Reviewer: @arkodg See https://github.com/kubernetes-sigs/gateway-api/pull/2759#discussion_r1483295712
```suggestion
    // +kubebuilder:validation:MinItems=1
    SubjectAltNames []string `json:"subjectAltNames,omitempty"`
```

Author: afaik no implementation make SAN mandatory (only CA Cert is mandatory), curious why we want to do it here ?

Reviewer: Maybe I'm mistaken, but IIUC we have `// +optional` which says the field is not required, but `MinItems=1` means that if the field is set then there must be at least 1 entry.

Reviewer: @arkodg I tried it (`+optional` and `MinItems=1`) and it works as I said above, i.e., SAN is not mandatory.

Author: awesome thanks for checking, will update

# GEP-91: Client Certificate Validation for TLS terminating at the Gateway Listener

* Issue: [#91](https://github.com/kubernetes-sigs/gateway-api/issues/91)
* Status: Implementable 

(See definitions in [GEP Status][/contributing/gep#status].)

## TLDR

This GEP proposes a way to validate the TLS certificate presented by the downstream client to the server
(Gateway Listener in this case) during a [TLS Handshake Protocol][].

## Goals

> - Define an API field to specify the CA Certificate within the Gateway Listener configuration that can be used as a trust anchor to validate the certificates presented by the client. This use case has been been highlighted in the [Gateway API TLS Use Cases][] document under point 7.

Reviewer 2: TLS Use cases doc is
https://docs.google.com/document/d/17sctu2uMJtHmJTGtBi_awGB0YzoCLodtR6rUNmKMCs8/edit?usp=sharing

Author: yeah already linked on L125

This table highlights the support. Please feel free to add any missing implementations not mentioned below.

| Implementation | Support       |
|----------------|------------|
| Apache APISIX  | [ApisixTls.Client.CASecret](https://apisix.apache.org/docs/ingress-controller/tutorials/mtls/#mutual-authentication)      |
| Contour        | [HTTPProxy.Spec.VirtualHost.Tls.ClientValidation.CASecret](https://projectcontour.io/docs/v1.17.1/config/tls-termination/)      |
| Emissary Ingress| [TlSContext.Spec.Secret](https://www.getambassador.io/docs/emissary/latest/topics/running/tls/mtls)     |
| Gloo Edge      | [VirtualService.Spec.SSLConfig.SecretRef](https://docs.solo.io/gloo-edge/latest/guides/security/tls/server_tls/#configuring-downstream-mtls-in-a-virtual-service)      |
| Istio          | [Gateway.Spec.Servers.TLS.Mode](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway)      |
| Kong           | [mTLS Plugin](https://docs.konghq.com/hub/kong-inc/mtls-auth/)      |
| Traefik        | [TLSOption.Spec.ClientAuth](https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls)    |

### API

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Author: @robscott the context here is `listener` && `tls` , so `clientValidation` shouldn't confuse the user with upstream TLS naming imo
@danehans what are the advantages of the top level `client` field here ? do you see more knobs being added to `client` . This location seems like the wrong home for Gateway Certs (needed for upstream TLS)

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Reviewer 3: I agree that upstream TLS should stay as-is, e.g. `gateway.spec.listeners[].tls.certificateRefs`. If we wanted to set allowed ALPN protocols or a downstream client TLS connection timeout would that be considered client validation?

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Author: ah I see where you're going with this, knobs like `alpn` should live under `gateway.spec.listeners[].tls`
here's an example of how we implemented in Envoy Gateway as a field within a policy
https://gateway.envoyproxy.io/latest/api/extension_types/#tlssettings

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Reviewer 2: Yeah, I think that ideally, we should be structuring `tls` stanzas similarly everywhere, something like this:

```yaml
tls:
  server:
    <existing TLS settings>
  client:
    validation:
      <some settings>
    othersetting:
```

Gateway is GA, so we can't change it now, but adding the `client` stanza in would really help.

Doing something like this now also means that we can move `BackendTLSPolicy` in that direction too - it's still experimental and alpha, so we can change things if necessary.

Additionally, this means that we could conceivably add a top-level `tls` stanza to Gateway that could hold Gateway-wide TLS settings (like, for example, a Gateway-wide TLS client certificate).

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Author: cool, I'll go ahead and make this change

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Author: ive gone ahead and made this change, @youngnick would be great if you can elaborate on which fields will call  `othersettings` their home

> * Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection

Author: we touched upon this in the call today, `client.Validation` is mainly being added to create a similar look and feel with the `BackendTLSPolicy` API

to the Gateway.
* Introduce a `caCertificateRefs` field within `ClientValidationContext` that can be used to specify a list of CA Certificates that
can be used as a trust anchor to validate the certificates presented by the client.
* This new field is mutually exclusive with the [BackendTLSPolicy][] configuation which is used to validate the TLS certificate presented by the peer on the connection between the Gateway and the backend, and this GEP is adding support for validating the TLS certificate presented by the peer on the connection between the Gateway and the downstream client.
* Introduce an optional `subjectAltNames` field within `ClientValidationContext` that can be used to specify one or more alternate names to verify the subject identity in the certificate presented by the client. The maximum number of alternate names that can be specified is implementation defined.

#### GO

```go
// ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection
// to the Gateway.
// By default, no client specific configuration is validated.

> type ClientValidationContext struct {

Author: good point, I'd prefer if permissive mode is introduced later as an enhancement into this API

> type ClientValidationContext struct {

Reviewer 4: Please don't use the word 'permissive' - Istio has created a lot of confusion and mess on that area, really means 'allow insecure'.

Also there is nothing 'untrusted' about requests without client certs - JWTs are equivalent, and many systems allow multiple form of authentication. JWT bound to client certs are another form - that is different from client certs.

A section with the APIs used by Nginx, apache, etc would be helpful to get the terms right - I suspect what we're talking about is the 'optional' config, where server ask but client may not send client cert and use other forms of auth.

> type ClientValidationContext struct {

Author: cool, I'll update to be more specific - "support optional certificate validation" in the future

    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of
    // the Certificate Authorities that can be used
    // as a trust anchor to validate the certificates presented by the client.
    //
    // A single CA certificate reference to a Kubernetes ConfigMap or Secret kind

>     // has "Core" support.

Author: hey @robscott is this a blocking comment ?

>     // has "Core" support.

Reviewer 3: xref cert-manager secret ref: https://cert-manager.io/docs/usage/certificate/#target-secret

> The signed certificate and private key are stored in the specified Secret resource.

Although it might seem obvious, should the spec define the type of Secret, e.g. `type: kubernetes.io/tls`?

Since the field supports configmap or secret references, we should consider defining an equivalent type, e.g. `SecretOrConfigMapObjectReference`.

>     // has "Core" support.

Reviewer 4: Isn't kubernetes.io/tls supposed to hold TLS private key and certificate ? 

IMO it is quite confusing to use Secret for public info like the root CAs to trust.

How about 'SYSTEM' ( i.e. client certs signed by the public roots of trust - many do sign email certs) ? 
Any prior art from kubeadmin or other k8s APIs dealing with setting the public certificates ? I know many support in-line certificates ( all webhooks ). Would be good to be consistent.

>     // has "Core" support.

Author: I'm moving this back to only supporting `ConfigMap` as Core, this doesnt block implementations (such as Istio and Envoy Gateway) from also supporting Secret

    // a Listener, but this behavior is implementation-specific.
    //
    // Support: Core - An optional single reference to a single Kubernetes Secret
    // and to a single Kubernetes ConfigMap with the CA certificate in a key named `ca.crt`.
    //
    // Support: Implementation-specific (More than one reference, or other kinds
    // of resources).
    //
    // References to a resource in a different namespace are invalid UNLESS there
    // is a ReferenceGrant in the target namespace that allows the certificate
    // to be attached. If a ReferenceGrant does not allow this reference, the
    // "ResolvedRefs" condition MUST be set to False for this listener with the
    // "RefNotPermitted" reason.
    //
    // +kubebuilder:validation:MaxItems=8
    // +kubebuilder:validation:MinItems=1
    CACertificateRefs []SecretObjectReference `json:”caCertificateRefs,omitempty”`
    // SubjectAltNames contains one or more alternate names to verify
    // the subject identity in the certificate presented by the client.
    //
    // Support: Core
    //
    // +optional
    // +kubebuilder:validation:MinItems=1

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Author: @robscott CN is deprecated , highlighted in this RFC https://datatracker.ietf.org/doc/html/rfc2818#section-3.1

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Author: pretty sure Envoy does not support CN validation https://github.com/envoyproxy/envoy/issues/4174, will need to cross check

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Author: regarding 2. is the API you were hoping for ?

```
subjectAltNames:
- type: "DNS"
   value: client.example.com
```

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Reviewer 5: @costinm will follow up on this

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Reviewer 2: Yes, the `type` field should stay, and should be optional and defaulted to `DNS`. It at least also needs `URI` as an option.

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Reviewer 4: To clarify: the deprecation is for CN use for _server_ identification, not client identification. 

Using DNS SAN for server identification is standard. Using DNS SAN for client identification is unusual - never saw it in use.

Email, URL SANs are used sometimes, different fields in the CN as well. I believe K8S client certs use the org or group.

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Reviewer 4: Also as mentioned above - due to size of the yaml file and cost to update this, it would be far better to keep authz separate.

>     SubjectAltNames []string `json:"subjectAltNames,omitempty"`

Author: removing SAN from this PR for now

}

```

#### YAML

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: client-validation-basic
spec:
  gatewayClassName: acme-lb
  listeners:
  - name: foo-https
    protocol: HTTPS
    port: 443
    hostname: foo.example.com
    tls:
      certificateRefs:
      - kind: Secret
        group: ""
        name: foo-example-com-cert
      clientValidation:
        caCertificateRefs:
        - kind: ConfigMap
          group: ""

>           name: foo-example-com-ca-cert

Reviewer 5: Have you considered making this a top level field instead of per listener? It's current placement requires SNI-matching prior to Client Cert validation which could be tricky. (I'm not necessarily advocating in one direction or the other here, just wanted to start a conversation).

Reviewer: I'm not sure I understand, but I think all the TLS handshake info should be together like this.

Anyway, this PR is just implementing what @arkodg is proposing for GEP-91 in #2273, so design discussions should probably be there, and I'll keep this PR in sync with any changes that happen in that one.

Author: @robscott implementations are already doing SNI matching today to enforce listener isolation, `hostname` is defined within each `listener`, so this works fine, this even works when Gateway API listeners needs to be flattened out in the implementation data plane in case 2 listeners have the same `port`.
I didn't quite understand what the  top level definition would enable ?

Reviewer 5: > I think all the TLS handshake info should be together like this.

Jumping forward a bit, I think it's very likely that the client cert a Gateway uses to connect to backends will be configured per Gateway (not per Listener) based on the discussion at the community meeting earlier this week. Essentially the argument is that in that case, the client cert is representing the identity of the Gateway, not any individual listener.

If we were to take that argument further, would a Gateway admin rather define the set of clients it trusts in one place or individually for each Listener? 

> implementations are already doing SNI matching today to enforce listener isolation

Some may end up with a subset of this. For example, many cloud LBs prevent domain fronting but don't expose full SNI matching config. Preventing domain fronting is actually one way to meet the requirements of SNI/hostname matching in Gateways, but that logic may not extend to things like client cert validation.

One detail that's not clear to me from the list of implementations that support client cert validation in this GEP is how many support configuring this differently per SNI on the same VIP + Port. That would help show the viability of this. For what it's worth, in GCP this config is per VIP + Port, not per SNI (https://cloud.google.com/load-balancing/docs/mtls). Of course a single implementation does not need to impact the entire design here, my hunch is that GCP may not be the only one with this limitation.

Author: resolving this comment since its being addressed by Rob with https://github.com/kubernetes-sigs/gateway-api/pull/2910

# GEP-91: Client Certificate Validation for TLS terminating at the Gateway Listener

* Issue: [#91](https://github.com/kubernetes-sigs/gateway-api/issues/91)
* Status: Implementable 

(See definitions in [GEP Status][/contributing/gep#status].)

## TLDR

This GEP proposes a way to validate the TLS certificate presented by the downstream client to the server
(Gateway Listener in this case) during a [TLS Handshake Protocol][].

## Goals

> - Define an API field to specify the CA Certificate within the Gateway Listener configuration that can be used as a trust anchor to validate the certificates presented by the client. This use case has been been highlighted in the [Gateway API TLS Use Cases][] document under point 7.

Reviewer 6: ```suggestion
- Define an API field to specify the CA Certificate within the Gateway Listener configuration that can be used as a trust anchor to validate the certificates presented by the client. This use case has been been highlighted in the [Gateway API TLS Use Cases] document under point 7.
```

?

Reviewer 5: Note that both the original and suggestion have "been been", suggest removing one of those. I also generally prefer inline links to aliases to avoid confusion, unless a link is going to be reused a lot.

Author: thanks fixed the typo

| Gloo Edge      | [VirtualService.Spec.SSLConfig.SecretRef](https://docs.solo.io/gloo-edge/latest/guides/security/tls/server_tls/#configuring-downstream-mtls-in-a-virtual-service)      |
| Istio          | [Gateway.Spec.Servers.TLS.Mode](https://istio.io/latest/docs/tasks/traffic-management/ingress/secure-ingress/#configure-a-mutual-tls-ingress-gateway)      |
| Kong           | [mTLS Plugin](https://docs.konghq.com/hub/kong-inc/mtls-auth/)      |
| Traefik        | [TLSOption.Spec.ClientAuth](https://doc.traefik.io/traefik/https/tls/#client-authentication-mtls)    |

### API

* Introduce a `Client` field of type `TLSClientContext` within [GatewayTLSConfig][] to hold TLS configuration specific to the client.
* Introduce a `Validation` field of type `TLSValidationContext` within the `Client` field that can be used to validate the peer with which the TLS connection is being made.
to the Gateway.
* Introduce a `caCertificateRefs` field within `ValidationContext` that can be used to specify a list of CA Certificates that
can be used as a trust anchor to validate the certificates presented by the client.

> * This new field is mutually exclusive with the [BackendTLSPolicy][] configuation which is used to validate the TLS certificate presented by the peer on the connection between the Gateway and the backend, and this GEP is adding support for validating the TLS certificate presented by the peer on the connection between the Gateway and the downstream client.

Reviewer 7: Would you mind clarifying "mutually exclusive"?  Let's say my use case is to have a Gateway that includes tls.client.validation.caCertificateRefs *and* has a backend Service with an attached BackendTLSPolicy and tls.caCertRefs. I'm indicating my trust for both the upstream connection (browser client cert) and downstream connection (backend server cert).  This use case - using both - is valid, right?

Author: yes, using both at the same time is valid, and mutually exclusive, let me update to make this clearer

> * Introduce an optional `subjectAltNames` field within `ClientValidationContext` that can be used to specify one or more alternate names to verify the subject identity in the certificate presented by the client. The maximum number of alternate names that can be specified is implementation defined.

Reviewer 4: I am not sure about this one - this is really authorization policy, and having a list of identities in the gateway itself is perhaps the worse for a gateway ( scale, personas doing the config, etc).

Reviewer 4: In Istio that's part of Authorization policy.

Author: @frankbu and @howardjohn from Istio requested for adding subjectAltName

Reviewer 4: Well, Istio does have this field in istio Gateway - but we also have the authorization policy as a separate resource. 

I don't think this is the best UX for verifying client certificates - do we have any real world use case where client certificates are using SANs ( real internet clients - not mesh ), and real internet gateways authenticate by listing all the clients in the config file ? Even the old httpasswd was a separate config and didn't scale well...

Reviewer 3: @arkodg Without a solid use case, it's best to drop `subjectAltNames` from the API to move this PR forward.

Reviewer 5: @howardjohn or @frankbu can you provide more info on why you'd like SANs to be configurable at this level?

Reviewer: Good question about real world use cases. I was only requesting this feature because Istio supports it and we'd rather not add an Istio-specific gateway-api extension if we can avoid it. https://github.com/istio/istio/pull/48013.

@howardjohn?

Reviewer 8: I generally agree this probably makes sense to split out as a more general authz. However, there is no authz in gateway-api so its a bit tricky..

Generally any TLS usage ought to actually verify the identity in the certificates; rarely is it sufficient to just validate the issuing CA. So we would end up in a state where to properly use this feature, each vendor would need to have its own policy and users would need to use it, or be left in an insecure state.

Author: removing SAN for now, will add a comment in the GEP mentioning the reason - "it falls under the authz bucket so we can revisit this field once the project tackles authz as a whole"

Reviewer 9: Knative uses these features for the gateway=>backend. Since those requests can be handled by a revision instance or the activator component (when we scale up).

We don't want our activator component impersonating all revision instances and having to deal with so many certificates)

Reviewer 9: On the client=>gateway side wouldn't allowing subjectAltNames enable auth with certs?

eg. My internal corp site requires my machine's client cert to match the alternate names `*.internal.broadcom.net` or `*.internal.vmware.com` ?

#### GO

```go
// TLSClientContext holds configuration specific to the client initiating the TLS connection.

> type TLSClientContext struct {

Reviewer 4: Worth mentioning that adding this field will result in servers adding a 'certificate request' in the TLS server hello - which results in browsers asking for a cert in a dialog. This is pretty unfortunate UX for the browsers - many users are not aware of this. 

So best to use a separate hostname or port for the mtls if browsers are used.

Reviewer 3: @costinm is ^ the case even if `CACertificateRefs` refs a CA cert that resides as an anchor in the client's system trust bundle?

Reviewer 4: Yes. 

mTLS works by server asking the client for a cert in ServerHelo - if server doesn't ask, client won't send. And if server asks, a browser is required to show the cert selection dialog ( at least first time ? Haven't tried in a long time). 

Note that most servers have options to make client cert optional or required - both still ask for cert, but one will allow the request to proceed and use JWT or other forms of auth if client cert is not available. 

This should probably be covered in the 'prior art' and background section...

Author: thanks ! incorporated this in as a doc string

# GEP-91: Client Certificate Validation for TLS terminating at the Gateway Listener

* Issue: [#91](https://github.com/kubernetes-sigs/gateway-api/issues/91)

> * Status: Implementable

Reviewer 5: Before this formally moves to experimental, I think we need a corresponding metadata file for this GEP. Not going to block on this specific PR since it can easily be handled in a following.

    // Setting this field will require clients to send a client certificate
    // required for validation during the TLS handshake. In browsers this may result in a dialog appearing 
    // that requests a user to specify the client certificate.
    // The maximum depth of a certificate chain accepted in verification is Implementation specific.
    FrontendValidation *FrontendTLSValidation `json:"frontendValidation,omitempty"`
}

// FrontendTLSValidation holds configuration information that can be used to validate
// the frontend initiating the TLS connection
type FrontendTLSValidationContext struct {
    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of
    // the Certificate Authorities that can be used
    // as a trust anchor to validate the certificates presented by the client.
    //
    // A single CA certificate reference to a Kubernetes ConfigMap
    // has "Core" support.
    // Implementations MAY choose to support attaching multiple CA certificates to
    // a Listener, but this behavior is implementation-specific.
    //
    // Support: Core - An optional single reference to a single Kubernetes Secret

>     // and to a single Kubernetes ConfigMap with the CA certificate in a key named `ca.crt`.

Reviewer 5: Nit: I'm not sure we really want to consider Secrets core support, that would mean we'd need to have conformance tests for both ConfigMaps and Secrets when we really just want people to use ConfigMaps and eventually ClusterTrustBundles.

Author: ah yes, thanks for finding this, fixed, was a left over, which I forgot to clean up

## Non-Goals
- Define other fields that can be used to verify the client certificate such as the Certificate Hash or Subject Alt Name. 

### API

* Introduce a `clientValidation` field of type `ClientValidationContext` within [GatewayTLSConfig][] that can be used to validate the client initiating the TLS connection
to the Gateway.
* Introduce a `caCertificateRefs` field within `ClientValidationContext` that can be used to specify a list of CA Certificates that
can be used as a trust anchor to validate the certificates presented by the client.

#### GO

```go
// ClientValidationContext holds configuration that can be used to validate the client intiating the TLS connection
// to the Gateway.
// By default, no client specific configuration is validated.
type ClientValidationContext struct {
    // CACertificateRefs contains one or more references to
    // Kubernetes objects that contain TLS certificates of

>     // the Certificate Authorities that can be used to

Reviewer: ```suggestion
    // the Certificate Authorities that can be used
```

    // as a trust anchor to validate the certificates presented by the client.
    //

>     // A single CACertRef to a Kubernetes ConfigMap kind has "Core" support.

Reviewer: ```suggestion
    // A single CACertificateRefs to a Kubernetes ConfigMap kind has "Core" support.
```

    // Implementations MAY choose to support attaching multiple CA certificates to
    // a Listener, but this behavior is implementation-specific.
    //

>     // Support: Core - An optional single reference to a Kubernetes ConfigMap,

Reviewer 8: Should a Secret with ca.crt be core as well? its not uncommon for things like cert-manager etc to write these

Author: good point, we plan on adding support for it in Envoy Gateway, 

for now, its very close to the semantics of BackendTrafficPolicy (which defines a ConfigMap as Core)

adding another new `Core` constraint in this PR may slow down the completion of this PR imo, prefer
if its added incrementally. If the community gives a big thumbs up for this, happy to add it to this PR

Author: Added Secret support to Core

    // with the CA certificate in a key named `ca.crt`.
    //
    // Support: Implementation-specific (More than one reference, or other kinds
    // of resources).
    //

>     // References to a resource in different namespace are invalid UNLESS there

Reviewer: ```suggestion
    // References to resources in a different namespace are invalid UNLESS there
```

    // is a ReferenceGrant in the target namespace that allows the certificate
    // to be attached. If a ReferenceGrant does not allow this reference, the
    // "ResolvedRefs" condition MUST be set to False for this listener with the
    // "RefNotPermitted" reason.
    //
    // +kubebuilder:validation:MaxItems=8
    // +optional

>     CACertificateRefs []corev1.ObjectReference `json:”caCertificateRefs,omitempty”`

Reviewer: Do we really want to use `corev1.ObjectReference` here? Seems that it should be either `SecretObjectReference` (same as GatewayTLSConfig.CertificateRefs) or `ConfigMapObjectReference` (same as gep-1897 BackendTLSPolicyConfig).

Author: pros for using `corev1.ObjectReference`
* this reference can be a `ConfigMap` with a `ca.crt` key or a `Secret` with a `ca.crt` (which can be the same secret that also has `tls.crt` and `tls.key` used in `CertificateRefs` )
* reference can live in a different namespace and `ResourceGrant` can be used to authorize the reference (similar to `CertificateRefs` )
